core/orchestrator.go

```diff
@@ -485,14 +486,89 @@ func (rt *RemoteTranscoder) Transcode(fname string, profiles []ffmpeg.VideoProfi
 	return [][]byte{}, fmt.Errorf("Unknown error")
 }
 
-func NewRemoteTranscoder(n *LivepeerNode, stream net.Transcoder_RegisterTranscoderServer) *RemoteTranscoder {
+func NewRemoteTranscoder(n *LivepeerNode, stream net.Transcoder_RegisterTranscoderServer, capacity int) *RemoteTranscoder {
 	return &RemoteTranscoder{
-		node:   n,
-		stream: stream,
-		eof:    make(chan struct{}, 1),
+		node:     n,
+		stream:   stream,
+		eof:      make(chan struct{}, 1),
+		capacity: capacity,
 	}
 }
 
+func NewRemoteTranscoderManager() RemoteTranscoderManager {
+	return RemoteTranscoderManager{
+		remoteTranscoders: []*RemoteTranscoder{},
+		liveTranscoders:   map[net.Transcoder_RegisterTranscoderServer]*RemoteTranscoder{},
+		RTmutex:           &sync.Mutex{},
+	}
+}
+
+type RemoteTranscoderManager struct {
+	remoteTranscoders []*RemoteTranscoder
+	liveTranscoders   map[net.Transcoder_RegisterTranscoderServer]*RemoteTranscoder
+	RTmutex           *sync.Mutex
+}
+
+func (rtm *RemoteTranscoderManager) Register(transcoder *RemoteTranscoder) {
+	rtm.RTmutex.Lock()
+	defer rtm.RTmutex.Unlock()
+
+	rtm.liveTranscoders[transcoder.stream] = transcoder
+
+	for i := 0; i < transcoder.Capacity(); i++ {
+		rtm.remoteTranscoders = append(rtm.remoteTranscoders, transcoder)
+	}
+}
+
+func (rtm *RemoteTranscoderManager) Unregister(t *RemoteTranscoder) {
+	rtm.RTmutex.Lock()
+	delete(rtm.liveTranscoders, t.stream)
```

Comment:
@j0sh @angyangie 
Should we delete from `remoteTranscoders` too? If we not, then O will start hitting dead transcoder again, even if it known that it is dead. Or maybe we could add flag `dead` to `RemoteTranscoder`, so O will not try It again if it was already unregistered?

Reply:
In `SelectTranscoder` any transcoders not in the `liveTranscoders` table are permanently removed from `remoteTranscoders`, so we wouldn't actually hit any dead transcoders. That bit of logic was borrowed from multi-O where it was important to release the lock quickly. In this case, it is less important to release the lock quickly since unregisters are probably not in the critical path. `SelectTranscoder` is still in the critical path.

However I'm not sure what would be better: to hold the lock longer here in a (probably) non-critical path, and iterate through the entire list of `remoteTranscoders`, or do shorter iterations as-needed during `SelectTranscoder` ?

Reply:
Yeah, my mistake - somehow looking into logs I thought it hits dead transcoders.

>However I'm not sure what would be better

I think it is best how it is now - now, for dead transcoder, it will quickly cut out that transcoder from end of the list. If we will move it into `Unregister` - then we will be actually copying parts of the list, and it will probably will be slower.